### ISSUE TITLE
Model not converging with DDP

### ISSUE THREAD

**Issue by JuanPZuluaga** [speechbrain/speechbrain#765]
Hello, 

I'm training an ASR Transformer model for Librispeech but I'm not getting acceptable results (not converging at all after some epochs). I'm using pytorch.distributed.launch. 

I'm using around 20 GPUs. I have seen some issues about using DDP. Is there any recipe implementing it? 

thanks, 
J Pablo

**Comment by TParcollet** [comment]
Having such a high number of GPUs might induce changing the configuration file a bit (especially batch size and gradient accumulation). Could you please share your YAML with us?

**Comment by JuanPZuluaga** [comment]
The only changes I did so far in the YAML are:

```
batch_size: 4 # what the GPUs allow
grandient_accumulation: 2
```
So the effective batch_size would be 160 (20 GPUs),

Juan

**Comment by TParcollet** [comment]
@popcornell should we give the new scheduler here to see if it helps? Because the configuration doesn't seem to be wrong from what I can see.

**Comment by JuanPZuluaga** [comment]
Good to add, that I'm running my script like this:

``` 
python3 -m pytorch.distributed.launch
        --nproc_per_node=1 \
        --nnodes=20 \
        --node_rank=0 \
        --master_addr=${master_id} \
        --master_port=${master_port} \
        train.py config.yaml --distributed_launch --distributed_backend 'nccl'  
```

I don't have direct access to the GPUs. Normally, I need to 'make' a request to the grid (SunGrid) and as soon as I get the 'master address' (GPU), I launch the remaining N jobs-1. I think the configuration is done correctly because I can follow up on the train_log.txt file. And actually, the number of steps in the 'progress bar' is updated (I mean, more GPUs, fewer steps)...

**Comment by TParcollet** [comment]
@aheba can you check that this command is ok? @JuanPZuluaga Can you share the train_log.txt you got so far ?

**Comment by JuanPZuluaga** [comment]
This is what I've so far: 

```
epoch: 1, lr: 6.67e-06, steps: 732, optimizer: Adam - train loss: 3.64e+02 - valid loss: 1.61e+02, valid ACC: 7.62e-02
epoch: 2, lr: 1.34e-05, steps: 1464, optimizer: Adam - train loss: 2.51e+02 - valid loss: 1.49e+02, valid ACC: 1.27e-01
epoch: 3, lr: 2.00e-05, steps: 2196, optimizer: Adam - train loss: 2.37e+02 - valid loss: 1.44e+02, valid ACC: 1.43e-01
epoch: 4, lr: 2.67e-05, steps: 2928, optimizer: Adam - train loss: 2.31e+02 - valid loss: 1.41e+02, valid ACC: 1.64e-01
epoch: 5, lr: 4.01e-05, steps: 4393, optimizer: Adam - train loss: 2.27e+02 - valid loss: 1.39e+02, valid ACC: 1.62e-01
epoch: 6, lr: 5.35e-05, steps: 5858, optimizer: Adam - train loss: 2.22e+02 - valid loss: 1.38e+02, valid ACC: 1.66e-01
epoch: 7, lr: 6.68e-05, steps: 7323, optimizer: Adam - train loss: 2.20e+02 - valid loss: 1.37e+02, valid ACC: 1.74e-01
epoch: 8, lr: 8.02e-05, steps: 8788, optimizer: Adam - train loss: 2.17e+02 - valid loss: 1.35e+02, valid ACC: 1.83e-01
epoch: 9, lr: 9.36e-05, steps: 10253, optimizer: Adam - train loss: 2.16e+02 - valid loss: 1.34e+02, valid ACC: 1.88e-01
epoch: 10, lr: 1.07e-04, steps: 11718, optimizer: Adam - train loss: 2.14e+02 - valid loss: 1.33e+02, valid ACC: 1.93e-01, valid WER: 90.06
````

I can let it run for the 100+ epochs, but not sure about the convergence speed...

**Comment by popcornell** [comment]
> @popcornell should we give the new scheduler here to see if it helps? Because the configuration doesn't seem to be wrong from what I can see.

What I can do is to just have a separate pull-request for that one. It is super simple I can merge it like in the next hour or so.

**Comment by TParcollet** [comment]
Let's do it

**Comment by popcornell** [comment]
merged it https://github.com/speechbrain/speechbrain/pull/769

**Comment by JuanPZuluaga** [comment]
Thanks for the fix. 

Still, I have one more -dumb- question: how the DDP problem is impacted by the modified files? I just saw that the changes were in the LR scheduler. Could it be a scaling factor at the moment of AllReduce or so?

**Comment by TParcollet** [comment]
It's not only the DDP one. All models will converge faster with this scheme. Honestly, we (and others as well) observe weird behaviors when running with DDP. Please try this new scheduler.  If it still doesn't converge, we will try other things....

**Comment by popcornell** [comment]
Basically the lr schedule makes a night and day difference. Transformers are highly sensitive to lr and lr schedule (and batch size and gradient accumulation) especially in ASR. Probably using a different optimizer like RAdam can help reducing this sensitivity.

**Comment by JuanPZuluaga** [comment]
Hey guys! Thanks for the update. 

It did work. Now I got like 6.6 WER in epoch 20 (training with 24 GPUs). 
Yeah, I saw similar issues regarding the LR and the scheduler. I'm trying this fix now in a very noisy database... I'll keep you updated about this. Maybe I could try RAdam in the future... 

JuanP

**Comment by TParcollet** [comment]
COuld thanks. Please let us know how the 24GPUs will affect the performance in the end.

**Comment by JuanPZuluaga** [comment]
Hello, 

I got the results! They are worse :/: 

```
Epoch loaded: 110 - test loss: 1.46, test ACC: 9.36e-01, test WER: 9.86 (test-clean)
Epoch loaded: 110 - test loss: 2.04, test ACC: 8.98e-01, test WER: 18.80 (test-other)
```
I might try the transducers or seq-to-seq recipes.
